datashader/datatypes.py

```diff
@@ -0,0 +1,1018 @@
+                    item += len(self)
+
+                slice_start = self.start_indices[item]
+                slice_end = (self.start_indices[item+1]
+                             if item + 1 <= len(self) - 1
+                             else len(self.flat_array))
+
+                return (self.flat_array[slice_start:slice_end]
+                        if slice_end!=slice_start
+                        else np.nan)
+
+        elif type(item) == slice:
+            data = []
+            selected_indices = np.arange(len(self))[item]
+
+            for selected_index in selected_indices:
+                data.append(self[selected_index])
+
+            return RaggedArray(data, dtype=self.flat_array.dtype)
+
+        elif isinstance(item, np.ndarray) and item.dtype == 'bool':
+            data = []
+
+            for i, m in enumerate(item):
+                if m:
+                    data.append(self[i])
+
+            return RaggedArray(data, dtype=self.flat_array.dtype)
+        elif isinstance(item, (list, np.ndarray)):
+            return self.take(item, allow_fill=False)
+        else:
+            raise IndexError(item)
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        """
+        Construct a new RaggedArray from a sequence of scalars.
+
+        Parameters
+        ----------
+        scalars : Sequence
+            Each element will be an instance of the scalar type for this
+            array, ``cls.dtype.type``.
+        dtype : dtype, optional
+            Construct for this particular dtype. This should be a Dtype
+            compatible with the ExtensionArray.
+        copy : boolean, default False
+            If True, copy the underlying data.
+
+        Returns
+        -------
+        RaggedArray
+        """
+        return RaggedArray(scalars, dtype=dtype)
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        """
+        Reconstruct an ExtensionArray after factorization.
+
+        Parameters
+        ----------
+        values : ndarray
+            An integer ndarray with the factorized values.
+        original : RaggedArray
+            The original ExtensionArray that factorize was called on.
+
+        See Also
+        --------
+        pandas.factorize
+        ExtensionArray.factorize
+        """
+        return RaggedArray(
+            [_RaggedElement.array_or_nan(v) for v in values],
+            dtype=original.flat_array.dtype)
+
+    def _as_ragged_element_array(self):
+        return np.array([_RaggedElement.ragged_or_nan(self[i])
+                         for i in range(len(self))])
+
+    def _values_for_factorize(self):
+        return self._as_ragged_element_array(), np.nan
+
+    def _values_for_argsort(self):
+        return self._as_ragged_element_array()
+
+    def unique(self):
+        """
+        Compute the ExtensionArray of unique values.
+
+        Returns
+        -------
+        uniques : ExtensionArray
+        """
```

Comment:
I would prefer not to copy docstrings unmodified from the parent class, whether trivial (as here) or complicated (below).  Basically, if the parent class defines the semantics, I want the reader to refer to the parent class, not to this possibly outdated copy of the docstring; that way people know to go find it in the parent, rather than thinking this actually covers everything.  Conversely, if there is a docstring here, I think it should be customized to just be about RaggedArray.

Reply:
Removed in 1538909